code/modules/research/rdconsole.dm

```diff
@@ -147,417 +134,266 @@ Nothing else in the console has ID requirements.
-	var/list/l = list()
-	if (stored_research.hidden_nodes[node.id])
-		return l
-	var/display_name = node.display_name
-	if (selflink)
-		display_name = "<A href='?src=[REF(src)];view_node=[node.id];back_screen=[screen]'>[display_name]</A>"
-	l += "<div class='statusDisplay technode'><b>[display_name]</b> [RDSCREEN_NOBREAK]"
-	if(minimal)
-		l += "<br>[node.description]"
-	else
-		if(stored_research.researched_nodes[node.id])
-			l += "<span class='linkOff'>Researched</span>"
-		else if(stored_research.available_nodes[node.id])
-			if(stored_research.can_afford(node.get_price(stored_research)))
-				l += "<BR><A href='?src=[REF(src)];research_node=[node.id]'>[node.price_display(stored_research)]</A>"
+/obj/machinery/computer/rdconsole/ui_interact(mob/user, datum/tgui/ui = null)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "Techweb")
+		ui.open()
+
+/obj/machinery/computer/rdconsole/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/research_designs)
+	)
+
+// heavy data from this proc should be moved to static data when possible
+/obj/machinery/computer/rdconsole/ui_data(mob/user)
+	. = list(
+		"nodes" = list(),
+		"experiments" = list(),
+		"researched_designs" = stored_research.researched_designs,
+		"points" = stored_research.research_points,
+		"points_last_tick" = stored_research.last_bitcoins,
+		"web_org" = stored_research.organization,
+		"sec_protocols" = !(obj_flags & EMAGGED),
+		"t_disk" = null,
+		"d_disk" = null,
+		"locked" = locked
+	)
+
+	if (t_disk)
+		.["t_disk"] = list (
+			"stored_research" = t_disk.stored_research.researched_nodes
+		)
+	if (d_disk)
+		.["d_disk"] = list (
+			"max_blueprints" = d_disk.max_blueprints,
+			"blueprints" = list()
+		)
+		for (var/i in 1 to d_disk.max_blueprints)
+			if (d_disk.blueprints[i])
+				var/datum/design/D = d_disk.blueprints[i]
+				.["d_disk"]["blueprints"] += D.id
 			else
-				l += "<BR><span class='linkOff'>[node.price_display(stored_research)]</span>"  // gray - too expensive
-		else
-			l += "<BR><span class='linkOff bad'>[node.price_display(stored_research)]</span>"  // red - missing prereqs
-		if(ui_mode == RDCONSOLE_UI_MODE_NORMAL)
-			l += "[node.description]"
-			for(var/i in node.design_ids)
-				var/datum/design/D = SSresearch.techweb_design_by_id(i)
-				l += "<span data-tooltip='[D.name]' onclick='location=\"?src=[REF(src)];view_design=[i];back_screen=[screen]\"'>[D.icon_html(usr)]</span>[RDSCREEN_NOBREAK]"
-	l += "</div>[RDSCREEN_NOBREAK]"
-	return l
-
-/obj/machinery/computer/rdconsole/proc/ui_techweb_nodeview()
-	var/datum/techweb_node/selected_node = SSresearch.techweb_node_by_id(selected_node_id)
-	RDSCREEN_UI_SNODE_CHECK
-	var/list/l = list()
-	if(stored_research.hidden_nodes[selected_node.id])
-		l += "<div><h3>ERROR: RESEARCH NODE UNKNOWN.</h3></div>"
+				.["d_disk"]["blueprints"] += null
```

Comment:
What's the point of the if statement if you're just going to `null` anyway?

Replace this whole thing with:

```dm
 		for (var/i in 1 to d_disk.max_blueprints)
			var/datum/design/blueprint = design_disk.blueprints[i]
			.["d_disk"]["blueprints"] += blueprint?.id
```


(Note the better variable name of d)